runtime/lua/vim/lsp/handlers.lua

```diff
@@ -512,6 +512,64 @@ M['window/showMessage'] = function(_, result, ctx, _)
   return result
 end
 
+--see: https://microsoft.github.io/language-server-protocol/specifications/specification-current/#window_showDocument
+M['window/showDocument'] = function(_, result, ctx, _)
+  local uri = result.uri
+
+  if result.external then
+    local cmd
+    if vim.fn.has('win32') == 1 then
+      cmd = { 'cmd.exe', '/c', 'start', '""', vim.fn.shellescape(uri) }
+    else
+      cmd = { 'xdg-open', vim.fn.shellescape(uri) }
+    end
+
+    local res = vim.fn.system(cmd)
+    if vim.v.shellerror ~= 0 then
+      return {
+        success = false,
+        error = {
+          code = protocol.ErrorCodes.UnknownErrorCode,
+          message = res,
+        },
+      }
+    end
+
+    return { success = true }
+  end
+
+  local client_id = ctx.client_id
+  local client = vim.lsp.get_client_by_id(client_id)
+  local client_name = client and client.name or string.format('id=%d', client_id)
+  if not client then
+    err_message({ 'LSP[', client_name, '] client has shut down after sending ', ctx.method })
+    return vim.NIL
+  end
+
+  local location = {
+    uri = uri,
+    range = result.selection or { start = { line = 0, character = 0 } },
+  }
+
+  local success
+  local set_cursor = result.selection and true or false
+
+  local prev = vim.api.nvim_get_current_win()
+  if not result.takeFocus then
+    local bufnr = vim.uri_to_bufnr(uri)
+    if vim.fn.bufwinid(bufnr) == -1 then
+      vim.cmd('new')
+    end
+  end
+
+  success = util.jump_to_location(location, client.offset_encoding, true, set_cursor)
```

Comment:
Could consider making `range` of `location` nullable as replacement for the `set_cursor` parameter.

Would keep the API smaller and make it more explicit that there was no range.

Reply:
That's way better. Thanks.